src/merkle/merkle_tree.rs

```diff
@@ -19,31 +20,84 @@ impl MerkleTree {
     ///
     /// # Errors
     /// Returns an error if the number of leaves is smaller than two or is not a power of two.
-    pub fn new(leaves: Vec<Word>) -> Result<Self, MerkleError> {
+    pub fn new<T, I, V>(leaves: T) -> Result<Self, MerkleError>
+    where
+        T: IntoIterator<IntoIter = I>,
+        I: ExactSizeIterator + Iterator<Item = V>,
+        V: Borrow<Word>,
```

Comment:
This was done to support the benchmarks requested by @bobbinth , the previous constructor was moving ownership of the vector the the MerkleTree. I had 2 options 1. create a new vector on every iteration of the benchmark, including that in the benchmark and making it harder to interpret the results 2. pass the values by reference and initialize the data only once.

I went with option two. `ExactSizeIterator` is required to figure out the len of the leaves and compute the storage to be used on `Vec::with_capacity`, `Borrow<Word>` is required because the type I got from criterion was `&Vec<&Word>`.

Reply:
I _think_ I found a solution. But it requires a nightly feature [`maybe_uninit_write_slice`](https://doc.rust-lang.org/core/mem/union.MaybeUninit.html#method.write_slice).

The code would then instead be:

```rust
    pub fn new<T>(leaves: T) -> Result<Self, MerkleError>
    where
        T: AsRef<[Word]>,
    {
       ...
       MaybeUninit::write_slice(last_layer, leaves);
```

Reply:
I pushed the code that uses the nightly here https://github.com/0xPolygonMiden/crypto/commit/6a6c9f8802968d99d4944091d57d1124b72324fa let me know if we should go with that